File: features/snippets/external_storage/custom_driver/custom_storage_driver.ts
import { randomUUID } from 'node:crypto';
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import path from 'node:path';
import {
  StorageDriverClaim,
  type Payload,
  type StorageDriver,
  type StorageDriverRetrieveContext,
  type StorageDriverStoreContext,
} from '@temporalio/common';
import { temporal } from '@temporalio/proto';

const PayloadProto = temporal.api.common.v1.Payload;

// @@@SNIPSTART typescript-custom-storage-driver
export class LocalDiskStorageDriver implements StorageDriver {
  readonly name = 'my-local-disk';
  readonly type = 'local-disk';

  constructor(private readonly storeDir = '/tmp/temporal-payload-store') {}

  async store(context: StorageDriverStoreContext, payloads: Payload[]): Promise<StorageDriverClaim[]> {
    let dir = this.storeDir;
    const target = context.target;
    if (target?.id) {
      // `target.kind` is either 'workflow' or 'activity'. Including it in the path keeps a
      // Workflow Id from colliding with an Activity Id in the same Namespace.
      dir = path.join(this.storeDir, target.namespace, target.kind, target.id);
    }
    await mkdir(dir, { recursive: true });

    const claims: StorageDriverClaim[] = [];
    for (const payload of payloads) {
      const filePath = path.join(dir, `${randomUUID()}.bin`);
      await writeFile(filePath, PayloadProto.encode(payload).finish());
      claims.push(new StorageDriverClaim({ path: filePath }));
    }
    return claims;
  }

  async retrieve(_context: StorageDriverRetrieveContext, claims: StorageDriverClaim[]): Promise<Payload[]> {
    const payloads: Payload[] = [];
    for (const claim of claims) {
      const filePath = claim.claimData.path;
      if (!filePath) {
        throw new Error("claim is missing required 'path' data");
      }
      payloads.push(PayloadProto.decode(await readFile(filePath)));
    }
    return payloads;
  }
}
// @@@SNIPEND

File: features/snippets/external_storage/custom_driver/custom_storage_driver.java
import io.temporal.api.common.v1.Payload;
import io.temporal.payload.storage.StorageDriver;
import io.temporal.payload.storage.StorageDriverActivityInfo;
import io.temporal.payload.storage.StorageDriverClaim;
import io.temporal.payload.storage.StorageDriverRetrieveContext;
import io.temporal.payload.storage.StorageDriverStoreContext;
import io.temporal.payload.storage.StorageDriverTargetInfo;
import io.temporal.payload.storage.StorageDriverWorkflowInfo;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

// @@@SNIPSTART java-custom-storage-driver
class LocalDiskStorageDriver implements StorageDriver {
  private static final String CLAIM_PATH = "path";

  private final Path storeDir;

  LocalDiskStorageDriver(Path storeDir) {
    this.storeDir = storeDir;
  }

  @Override
  public String getName() {
    return "local-disk";
  }

  @Override
  public String getType() {
    return "local-disk";
  }

  @Override
  public CompletableFuture<List<StorageDriverClaim>> store(
      StorageDriverStoreContext context, List<Payload> payloads) {
    try {
      context.getCancellationToken().throwIfCancellationRequested();
      Path directory = storeDirectory(context);
      Files.createDirectories(directory);

      List<StorageDriverClaim> claims = new ArrayList<>();
      for (Payload payload : payloads) {
        context.getCancellationToken().throwIfCancellationRequested();
        Path file = directory.resolve(UUID.randomUUID() + ".bin");
        Files.write(file, payload.toByteArray());
        claims.add(new StorageDriverClaim(Map.of(CLAIM_PATH, file.toString())));
      }
      return CompletableFuture.completedFuture(claims);
    } catch (IOException e) {
      return failedFuture(new IllegalStateException("Could not write Payload", e));
    }
  }

  @Override
  public CompletableFuture<List<Payload>> retrieve(
      StorageDriverRetrieveContext context, List<StorageDriverClaim> claims) {
    try {
      context.getCancellationToken().throwIfCancellationRequested();

      List<Payload> payloads = new ArrayList<>();
      for (StorageDriverClaim claim : claims) {
        context.getCancellationToken().throwIfCancellationRequested();
        Path file = Paths.get(claim.getClaimData().get(CLAIM_PATH));
        payloads.add(Payload.parseFrom(Files.readAllBytes(file)));
      }
      return CompletableFuture.completedFuture(payloads);
    } catch (IOException e) {
      return failedFuture(new IllegalStateException("Could not read Payload", e));
    }
  }

  private Path storeDirectory(StorageDriverStoreContext context) {
    StorageDriverTargetInfo target = context.getTarget();
    if (target instanceof StorageDriverWorkflowInfo) {
      StorageDriverWorkflowInfo workflow = (StorageDriverWorkflowInfo) target;
      if (workflow.getId() != null) {
        return storeDir.resolve(workflow.getNamespace()).resolve(workflow.getId());
      }
    } else if (target instanceof StorageDriverActivityInfo) {
      StorageDriverActivityInfo activity = (StorageDriverActivityInfo) target;
      if (activity.getId() != null) {
        return storeDir.resolve(activity.getNamespace()).resolve(activity.getId());
      }
    }
    return storeDir;
  }

  private static <T> CompletableFuture<T> failedFuture(Throwable error) {
    CompletableFuture<T> result = new CompletableFuture<>();
    result.completeExceptionally(error);
    return result;
  }
}
// @@@SNIPEND
